cmd/s3-layer.go

```diff
@@ -77,73 +78,68 @@ func s3ToObjectError(err error, params ...string) error {
 		}
 	case "XMinioInvalidObjectName":
 		err = ObjectNameInvalid{}
+	case "AccessDenied":
+		err = PrefixAccessDenied{
+			Bucket: bucket,
+			Object: object,
+		}
 	default:
-		fmt.Println("Undefined s3 object error:", bucket, object, minioErr.Code)
+		// should not be reached
+		err = fmt.Errorf("Undefined s3 object error: bucket=%s object=%s code=%s", bucket, object, minioErr.Code)
 	}
 
 	e.e = err
 	return e
 }
 
-type s3LayerConfig struct {
-	Location        string
-	Endpoint        string
-	AccessKeyID     string
-	SecretAccessKey string
-	UseSSL          bool
-}
-
-// S3Layer - Implements Object layer for S3 and Minio blob storage.
-type S3Layer struct {
-	s3LayerConfig
-
-	Client *minio.Core
+// s3Gateway - Implements gateway for S3 and Minio blob storage.
+type s3Gateway struct {
+	Client     *minio.Core
+	anonClient *minio.Core
 }
 
-func newS3Layer(endpoint, location, accessKey, secretKey string) (GatewayLayer, error) {
-	config := s3LayerConfig{
-		Endpoint:        endpoint,
-		Location:        location,
-		AccessKeyID:     accessKey,
-		SecretAccessKey: secretKey,
-		UseSSL:          true,
+// newS3Gateway returns s3 gatewaylayer
+func newS3Gateway(endpoint, accessKey, secretKey string) (GatewayLayer, error) {
+	// Initialize minio client object.
+	client, err := minio.NewCore(endpoint, accessKey, secretKey, true)
```

Comment:
since we need to allow for http and https, endpoint would be `http://host.com` so endpoint needs to be parsed to see if the scheme is http or https and accordingly call `minio.NewCore` with the isSecure option